Set Eigen3_DIR when activating a moose-libmesh env

milljm opened the pull request:
At the request of Cardinal, set a convenient variable so Cardinal users do not have to.

Closes #23069

Commit:
Set Eigen3_DIR when activating a moose-libmesh env

At the request of Cardinal, set a convenient variable so Cardinal
users do not have to.

Closes #23069

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1283080/) on c00342b wanted to post the following:

A change of the following file(s) triggered this check:

  conda/libmesh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

aprilnovak commented on conda/libmesh/build.sh:
```diff
@@ -66,12 +66,14 @@ make -j $CORES
 mkdir -p "${PREFIX}/etc/conda/activate.d" "${PREFIX}/etc/conda/deactivate.d"
 cat <<EOF > "${PREFIX}/etc/conda/activate.d/activate_${PKG_NAME}.sh"
 export LIBMESH_DIR=${PREFIX}/libmesh
+export Eigen3_DIR=${PREFIX}/libmesh/include/Eigen
```
Does there also need to be a copy of the `$CONDA_PREFIX_1/pkgs/eigen-3.4.0-h940c156_0/share/eigen3/cmake/*` files into this `Eigen3_DIR`?

milljm replied:
CONDA_PREFIX_1 does not always exist. But more importantly what you see here are internal `conda build` environment variables needed during package creation only. `${PREFIX}` gets replaced with a special `placeholder` tag, that Conda understands to replace _that_ with the installation location of your Conda implementation environments:

```bash
bash Mambaforge-Linux-x86_64.sh -b -p ~/mambaforge3
                                      ^^^^^^^^^^^^ this location can be anything and ends up being
                                                   the absolute path replacing 'placeholder'. e.g:

mamba create -n moose moose-libmesh

placeholder = /Users/you/mambaforge3/envs/moose/
```
It is this neat feature, that makes Conda relocatable and installable to any location you have write access (we love it, because we are able to tell folks they don't need sudo to install a MOOSE developers environment).

All that aside, are there additional environment variables I am forgetting about?

aprilnovak replied:
Perhaps I'm still misunderstanding something. I'm trying to express something @GiudGiud mentioned on Slack.

DAGMC will error unless it finds the following CMake files inside of `$Eigen3_DIR`:

```
(moose) {12:01}~/projects/cardinal:  ➭ ls $Eigen3_DIR
Eigen3Config.cmake        Eigen3ConfigVersion.cmake Eigen3Targets.cmake       UseEigen3.cmake
```

Will those files be found inside of `$Eigen3_DIR` (which conda will now define as you've described above)?

milljm replied:
These concerns were discussed in Slack, and resolved. The above changes are correct.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1283080/) on c00342b wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1283103/) on c00342b wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23070/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1283116/) on c00342b wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/23070/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/23070/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/23070/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/23070/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/23070/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/23070/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/23070/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/23070/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/23070/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/23070/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/23070/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/23070/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/23070/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/23070/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/23070/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/23070/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/23070/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/23070/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/23070/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/23070/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/23070/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/23070/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/23070/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/23070/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/23070/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/23070/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/23070/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/23070/coverage/xfem)

</details>



This comment will be updated on new commits.

loganharbour commented:
We should do an update + apptainter script fix in with this if this is to go in

Commit:
Add dep_tag_prefix remote container dep check

Requested fix from Logan to be included into this simple PR

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1284470/) on 5991458 wanted to post the following:

A change of the following file(s) triggered this check:

  conda/libmesh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1284470/) on 5991458 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1284507/) on 5991458 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `contact` coverage rate (required: 87.0%)
- Failed to generate `electromagnetics` coverage rate (required: 95.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 86.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 90.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 83.0%)
- Failed to generate `geochemistry` coverage rate (required: 96.0%)
- Failed to generate `heat_conduction` coverage rate (required: 87.0%)
- Failed to generate `level_set` coverage rate (required: 88.0%)
- Failed to generate `misc` coverage rate (required: 55.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `peridynamics` coverage rate (required: 78.0%)
- Failed to generate `phase_field` coverage rate (required: 89.0%)
- Failed to generate `porous_flow` coverage rate (required: 96.0%)
- Failed to generate `ray_tracing` coverage rate (required: 94.0%)
- Failed to generate `rdg` coverage rate (required: 65.0%)
- Failed to generate `reactor` coverage rate (required: 90.0%)
- Failed to generate `scalar_transport` coverage rate (required: 90.0%)
- Failed to generate `solid_properties` coverage rate (required: 84.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)
- Failed to generate `tensor_mechanics` coverage rate (required: 86.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 90.0%)
- Failed to generate `xfem` coverage rate (required: 80.0%)

loganharbour commented:
@roystgnr any interest in a libmesh update while we're upping envs?

roystgnr commented:
There's stuff to go in, but I'm afk most of the weekend.  ETA?

loganharbour commented:
I'm fine with Monday.

roystgnr committed:
libMesh updates in newsletter

roystgnr committed:
libMesh submodule update

* Added `Elem::orient()` method, to automatically reorder "flipped"
  elements that would otherwise have a negative Jacobian.
* Added `MeshTools::Modification::orient_elements()` method, to
  automatically reorder any and all "flipped" elements in an entire
  mesh.
* Changed `GhostingFunctor::map_type` definition, to shorten +
  optimize `SparsityPattern` construction code.  Typical matrix
  initialization speed is unchanged, common initialization cases are
  approximately 50% faster, and worst-case initialization benchmarks
  are 800% faster.
* libMesh now only prints its own error messages from exceptions which
  are not caught.  Higher-level code catching error exceptions can
  choose whether to print their messages or not.
* Mesh refinement support for Prism21 elements
* `SIDE_HIERARCHIC` finite element support for prisms, higher shape
  function orders on hexes
* Added `Pyramid18` geometry and `Lagrange` basis support
* Nemesis files now support named blocks, sidesets, and nodesets
* Added a flag to control whether nodal quadrature is allowed in
  pyramids
* Compatibility fixes for MFFD with PETSc 3.18.1
* Replaced deprecated-by-OSX `sprintf()` calls with proper C++
* Added `NodeElem` implementations for `contains_point()` and
  `close_to_point()`
* Added `PetscVector::pointwise_divide()`
* `Xdr` now accepts any `std::string_view` compatible comment
* Bug fixes for Abaqus I/O, Tri7 refinement, `L2_HIERARCHIC` shapes on
  Tri3, and an unnecessary header inclusion

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit 3f05f40f4a45060c1fd07281f8869a4667086291
+Subproject commit a476dcb36d78923b9d2ed13b937b388f00da77f4
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1289112/) on c010393 wanted to post the following:

A change of the following file(s) triggered this check:

  conda/libmesh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1289112/) on c010393 wanted to post the following:

A change of the following file(s) triggered this check:

  libmesh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The libmesh submodule or configuration was changed but the conda build config was not

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1289112/) on c010393 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

cticenhour reviewed (dismissed):
Looks good to me. Thanks for the newsletter entry!

> worst-case initialization benchmarks are 800% faster

Don't see that sort of speed increase very often! 😂

cticenhour commented:
@loganharbour seeing this on the conda build targets:

```
BUILD_ROOT/moose/conda/libmesh/: New Commit. Cleaning and/or Indexing remote channel
BUILD_ROOT/moose/conda/libmesh/: Cleaning osx-arm64 and then Indexing channel
WARNING: Your password has expired.
Password change required but no TTY available.

ERROR: exiting with code 1
```

See https://civet.inl.gov/job/1289114/ for more info.

roystgnr commented:
> Don't see that sort of speed increase very often!

It was a group effort.  By which I mean that @lindsayad [did 99% of the work](https://github.com/libMesh/libmesh/issues/3422) while I [offered constructive criticism](https://www.youtube.com/watch?v=NpYEJx7PkWE).

cticenhour committed:
Increment MPICH, libmesh-vtk, and PETSc packages

Also removes apptainer_generator.py from versioner requirements

cticenhour committed:
Generate versioner hashes

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1289799/) on 1e99ec9 wanted to post the following:

A change of the following file(s) triggered this check:

  conda/petsc

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1289799/) on 1e99ec9 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1289799/) on 1e99ec9 wanted to post the following:

A change of the following file(s) triggered this check:

  libmesh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml

The libmesh submodule or configuration was changed but the conda build config was not

lindsayad commented:
> t was a group effort. By which I mean that @lindsayad [did 99% of the work](https://github.com/libMesh/libmesh/issues/3422) while I [offered constructive criticism](https://www.youtube.com/watch?v=NpYEJx7PkWE).

I think my memory is continually deteriorating. I had to remind myself what this was and it only happened two months ago
